circuit/src/gadgets/sum.rs

```diff
@@ -0,0 +1,226 @@
+					sum = region.assign_advice(|| "sum", config.sum, i + 1, || val)?;
+				}
+
+				Ok(sum)
+			},
+		)
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use std::usize;
+
+	use super::*;
+	use crate::utils::{generate_params, prove_and_verify};
+	use halo2wrong::{
+		curves::bn256::{Bn256, Fr},
+		halo2::{
+			arithmetic::Field,
+			circuit::{SimpleFloorPlanner, Value},
+			dev::MockProver,
+			plonk::{Circuit, Instance},
+		},
+	};
+
+	#[derive(Clone)]
+	struct TestConfig {
+		sum: SumConfig,
+		temp: Column<Advice>,
+		pub_ins: Column<Instance>,
+	}
+
+	#[derive(Clone)]
+	struct TestCircuit<F: FieldExt, const S: usize> {
+		items: [F; S],
+	}
+
+	impl<F: FieldExt, const S: usize> TestCircuit<F, S> {
+		fn new(items: [F; S]) -> Self {
+			Self { items }
+		}
+	}
+
+	impl<F: FieldExt, const S: usize> Circuit<F> for TestCircuit<F, S> {
+		type Config = TestConfig;
+		type FloorPlanner = SimpleFloorPlanner;
+
+		fn without_witnesses(&self) -> Self {
+			self.clone()
+		}
+
+		fn configure(meta: &mut ConstraintSystem<F>) -> TestConfig {
+			let sum = SumChip::<_, S>::configure(meta);
+			let temp = meta.advice_column();
+			let pub_ins = meta.instance_column();
+
+			meta.enable_equality(temp);
+			meta.enable_equality(pub_ins);
+
+			TestConfig { sum, temp, pub_ins }
+		}
+
+		fn synthesize(
+			&self,
+			config: TestConfig,
+			mut layouter: impl Layouter<F>,
+		) -> Result<(), Error> {
+			let arr = layouter.assign_region(
+				|| "temp",
+				|mut region: Region<'_, F>| {
+					let mut arr: [Option<AssignedCell<F, F>>; S] = [(); S].map(|_| None);
+					for i in 0..S {
+						arr[i] = Some(region.assign_advice(
+							|| "temp",
+							config.temp,
+							i,
+							|| Value::known(self.items[i]),
+						)?);
+					}
+					Ok(arr.map(|a| a.unwrap()))
+				},
+			)?;
+			let sum_chip = SumChip::new(arr);
+			let acc = sum_chip.synthesize(config.sum, layouter.namespace(|| "sum"))?;
+
+			layouter.constrain_instance(acc.cell(), config.pub_ins, 0)?;
+			Ok(())
+		}
+	}
+
+	#[test]
+	fn test_sum() {
+		// Testing three 2's representation on the field as an input.
+		let test_chip = TestCircuit::new([Fr::from(2); 3]);
+
+		let k = 4;
+		let pub_ins = vec![Fr::from(6)];
+		let prover = MockProver::run(k, &test_chip, vec![pub_ins]).unwrap();
+		assert_eq!(prover.verify(), Ok(()));
+	}
+
+	#[test]
+	fn test_sum_random_element() {
+		// Testing big array with a random element from the field.
+		let rng = &mut rand::thread_rng();
+		let array = [<Fr as Field>::random(rng); 20];
+		let test_chip = TestCircuit::new(array);
+
+		let k = 5;
+		let ins = vec![array[0].mul(&Fr::from(20))];
+		let prover = MockProver::run(k, &test_chip, vec![ins]).unwrap();
+		assert_eq!(prover.verify(), Ok(()));
+	}
+
+	#[test]
+	fn test_sum_zero_arr() {
+		// Testing zero array. Returns InvalidInstance error.
+		let test_chip: TestCircuit<Fr, 0> = TestCircuit::new([]);
+
+		let k = 4;
+		let prover = MockProver::run(k, &test_chip, vec![]);
```

Comment:
`InvalidInstance` error is given because the instance cell is not populated.
What we want is to test if empty array yields to sum of zero.
So perhaps pass `Fr::zero()` as a public input.

Reply:
I see that an empty array doesn't yield to sum of zero. It gives a different error which is because the chip is not satisfied. Are we aiming to get a result from the circuit when we give an empty array?

Reply:
Yes, there is probably no use-case for it, but we should cover all the edge-cases.

HINT: The selector is enabled at index 0, before we start the loop: https://github.com/lazovicff/eigen-trust/blob/sum-circuit-test-doc/circuit/src/gadgets/sum.rs#L75, maybe that's unnecessary

Reply:
Oh, okay. If selector never enables it will provide zero because of the initial sum will provide zero as a constant at the beginning. So we won't get an error about chip satisfaction.